### PR TITLE
Management API Second Iteration

### PR DESCRIPTION
**Pull request by SanojPunchihewa**

### CODE REVIEW

File: cmd/README.md [hunk 1]
```diff
@@ -10,28 +10,21 @@ Command Line tool for managing the WSO2 Micro Integrator
...
-   
-    Execute `./build.sh -t micli.go -v 1.0.0 -f` to build for all platforms. (Can only be built in Mac or Linux)
-      
-    Created packages will be available at `build/target` directory
+    5. `cd` into cloned directory and then cd into `micro-integrator`
+    6. Execute `mvn clean install` to build the project
     
 - ### Running
-    Extract the compressed archive generated to a desired location.
+    `cd` into `micro-integrator/cmd/bin`
```
Comment (a5anka) [inline cmd/README.md]: Do we have to change directory to execute the CLI? Can we run from outside?

File: cmd/cmd/endpointInfo.go [hunk 2]
```diff
@@ -45,134 +42,56 @@ Example:
...
 // printEndpointInfo
 // @param Endpoint : Endpoint object
 func printEndpoint(endpoint utils.Endpoint) {
 
-	table := tablewriter.NewWriter(os.Stdout)
+    table := tablewriter.NewWriter(os.Stdout)
 
-	d_name := []string{"NAME", endpoint.Name}
-	table.Append(d_name)
+    d_name := []string{"NAME", endpoint.Name}
```
Comment (a5anka) [inline cmd/cmd/endpointInfo.go]: Are we using the correct naming convention for variables? AFAIK normally we use camelcase.

Comment (a5anka) [reply cmd/cmd/endpointInfo.go]: In `endpointList.go` you have used camelcase. Shall we use one convention?

**Comment by a5anka** [comment]
What is out plan for testing these code? How do we make sure that we do not break the functionality implemented in this code?

Shall we create a GitHub issue if we cannot solve this immediately?

File: cmd/cmd/taskInfo.go [hunk 3]
```diff
@@ -44,138 +41,65 @@ Example:
...
-            utils.HandleErrorAndExit(utils.LogPrefixError+"invalid XML response", unmarshalError)
-        }
-
-        return taskResponse, nil
-    } else {
-        return nil, errors.New(resp.Status())
-    }
 }
 
 // printTaskInfo
```
Comment (a5anka) [inline cmd/cmd/taskInfo.go]: Shall we have meaningful comments instead of just having the method name?

File: cmd/pom.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    <parent>
+        <groupId>org.wso2.ei</groupId>
+        <artifactId>wso2-micro-integrator-parent</artifactId>
+        <version>1.0.0-m4-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>micli</artifactId>
+    <packaging>mvn-golang</packaging>
+
```
Comment (a5anka) [inline cmd/pom.xml]: Shall we add a `<name>` tag to the pom?

File: cmd/pom.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    <properties>
+        <mvn.golang.skip>true</mvn.golang.skip>
+        <maven.install.skip>true</maven.install.skip>
+        <maven.deploy.skip>true</maven.deploy.skip>
```
Comment (a5anka) [inline cmd/pom.xml]: Why do we have to set these properties when we use a separate profile?

File: cmd/pom.xml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    <build>
+        <sourceDirectory>${basedir}</sourceDirectory>
+        <directory>${basedir}${file.separator}bin</directory>
+
+        <plugins>
+
+            <plugin>
+                <groupId>com.igormaznitsa</groupId>
+                <artifactId>mvn-golang-wrapper</artifactId>
+                <version>2.3.1</version>
```
Comment (a5anka) [inline cmd/pom.xml]: We should not hardcode plugin versions in child poms. Shall we take the version from parent pom?

File: cmd/README.md (hunk 1 above)
Comment (SanojPunchihewa) [reply cmd/README.md]: To execute the CLI from anywhere, you need to append the location of the executable to your $PATH variable

Comment (a5anka) [reply cmd/README.md]: That's good. Shall we instruct the user to do that? Then it will look cleaner. WDYT?

Comment (SanojPunchihewa) [reply cmd/README.md]: That's better, I will update the README

File: distribution/src/conf/internal-apis.xml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied. See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+<apis>
+  <api name="ManagementInternalAPI" class="org.wso2.carbon.micro.integrator.management.apis.ManagementInternalAPI"/>
```
Comment (a5anka) [inline distribution/src/conf/internal-apis.xml]: What is the purpose of this configuration?

File: cmd/cmd/apiInfo.go [hunk 8]
```diff
@@ -47,119 +43,48 @@ Example:
...
-    //  } else {
-    //      // token only, proceed with token
-    //  }
-    // } else {
-    //  // no token provided with --token (-t) flag
-    //  // proceed with username and password
-    //  accessToken, apiManagerEndpoint, preCommandErr := utils.ExecutePreCommand(listApisCmdEnvironment, listApisCmdUsername,
-    //      listApisCmdPassword, utils.MainConfigFilePath, utils.EnvKeysAllFilePath)
-
-    //  if preCommandErr == nil {
```
Comment (Nirothipan) [inline cmd/cmd/apiInfo.go]: Shall we remove these commented code

File: cmd/cmd/apiInfo.go [hunk 9]
```diff
@@ -38,149 +33,136 @@ var apiName string
...
-		executeGetAPICmd(apiName)
-	},
+    Use:   showAPICmdLiteral,
+    Short: showAPICmdShortDesc,
+    Long:  showAPICmdLongDesc + showAPICmdExamples,
+    Run: func(cmd *cobra.Command, args []string) {
+        utils.Logln(utils.LogPrefixInfo + "Show API called")
+        if len(args) == 0 {
+            executeListAPIsCmd()
+        }else if len(args) == 1{
```
Comment (a5anka) [inline cmd/cmd/apiInfo.go]: Don't we need to have a space after the curly braces? Let's fix it in other places as well.

File: cmd/utils/errorHandling.go [hunk 10]
```diff
@@ -1,51 +1,53 @@
...
-	fmt.Println("=======  END OF ERROR LOG ===========")
-	if err == nil {
-		fmt.Fprintf(os.Stderr, "%s: %v\n", ProjectName, msg)
-	} else {
-		fmt.Fprintf(os.Stderr, "%s: %v Reason: %v\n", ProjectName, msg, err.Error())
-		Logln(LogPrefixError + msg + ": " + err.Error())
-	}
+    if IsVerbose {
+        fmt.Println("\n=======  ERROR LOG ==================")
+        // TODO:: Remove debug log in production
```
Comment (a5anka) [inline cmd/utils/errorHandling.go]: We should not have code with TODOs. Ideally, we should fix it or have a GitHub issue for it. Shall we do that?

File: components/org.wso2.carbon.micro.integrator.extensions/org.wso2.carbon.micro.integrator.management.apis/pom.xml [hunk 11]
```diff
@@ -47,4 +34,25 @@
...
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.carbon.micro.integrator.management.apis.*;version="${project.version}"
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (a5anka) [inline components/org.wso2.carbon.micro.integrator.extensions/org.wso2.carbon.micro.integrator.management.apis/pom.xml]: Do we really need Dynamix imports for this?

File: features/org.wso2.carbon.micro.integrator.server.feature/pom.xml [hunk 12]
```diff
@@ -223,6 +223,10 @@
                                 <bundleDef>org.wso2.carbon:org.wso2.carbon.addressing:${carbon.kernel.version}
                                 </bundleDef>
 
+                                <!--Management API-->
+
+                                <bundleDef>org.wso2.ei:org.wso2.carbon.micro.integrator.management.apis:${product.ei.version}</bundleDef>
```
Comment (a5anka) [inline features/org.wso2.carbon.micro.integrator.server.feature/pom.xml]: Do we need this much new lines spaces here?

Comment (SanojPunchihewa) [reply features/org.wso2.carbon.micro.integrator.server.feature/pom.xml]: Fixed in 1a041a80d14e7cfd3b221ddbcc9866fba9aafe0d

File: components/org.wso2.carbon.micro.integrator.extensions/org.wso2.carbon.micro.integrator.management.apis/pom.xml (hunk 11 above)
Comment (SanojPunchihewa) [reply components/org.wso2.carbon.micro.integrator.extensions/org.wso2.carbon.micro.integrator.management.apis/pom.xml]: Removed in 013ad42

File: cmd/utils/errorHandling.go (hunk 10 above)
Comment (SanojPunchihewa) [reply cmd/utils/errorHandling.go]: Removed debug log in 951ef48

File: pom.xml [hunk 13]
```diff
@@ -561,5 +561,4 @@
             </plugins>
         </pluginManagement>
     </build>
-</project>
-
+</project>
```
Comment (a5anka) [inline pom.xml]: Let's not remove the new line

File: features/org.wso2.carbon.micro.integrator.server.feature/pom.xml (hunk 12 above)
Comment (a5anka) [reply features/org.wso2.carbon.micro.integrator.server.feature/pom.xml]: It seems the spaces are back

File: cmd/README.md [hunk 14]
```diff
@@ -12,26 +12,32 @@ Command Line tool for managing the WSO2 Micro Integrator
...
-- ### Inbound Listening Port
-    The default listening port of the Management API is `9091`
+
+- ### How to Enable the Management API
+    By default the Management API is disabled. To use the Management API you must use the system property `-DenableManagementInternalAPI` when starting the micro integrator
+
+- ### Management API Address and Port
+    To configure the address and the port of the Management Api in the CLI use the [**init**](#init) command. This will generate a file called server_config.yaml which contains the address and the port. If the init command was not used, the address and the port will have the default values
+
+    NOTE: The default address is http://localhost and the port is 9191
```
Comment (Nirothipan) [inline cmd/README.md]: The port has to be 9201 since we have a default offset of 10 in MI.

File: cmd/utils/constants.go [hunk 15]
```diff
@@ -81,11 +81,13 @@ const DefaultHttpRequestTimeout = 100000
 
 // DO NOT CHANGE THESE MANUALLY
 // Default Server Address
-const Host = "http://localhost"
-const Port = "9091"
+const HTTPProtocol = "http://"
+const HTTPSProtocol = "https://"
+const DefaultHost = "localhost"
+const DefaultPort = "9191"
```
Comment (Nirothipan) [inline cmd/utils/constants.go]: Don't we need to consider port offset here ?

File: cmd/README.md [hunk 16]
```diff
@@ -12,26 +12,32 @@ Command Line tool for managing the WSO2 Micro Integrator
...
-    
+
     NOTE: To execute the tool from anywhere, append the location of the executable (micli) to your $PATH variable.
 
 ### Configuration 
-- ### Inbound Listening Port
-    The default listening port of the Management API is `9091`
+
+- ### How to Enable the Management API
+    By default the Management API is disabled. To use the Management API you must use the system property `-DenableManagementInternalAPI` when starting the micro integrator
```
Comment (Nirothipan) [inline cmd/README.md]: This has to be -DenableManagement**Api**  ( enable + <Internal_API_Name> )

File: cmd/README.md (hunk 14 above)
Comment (SanojPunchihewa) [reply cmd/README.md]: Fixed in 18ec375

File: cmd/utils/constants.go (hunk 15 above)
Comment (SanojPunchihewa) [reply cmd/utils/constants.go]: Fixed in 18ec375

File: cmd/README.md (hunk 16 above)
Comment (SanojPunchihewa) [reply cmd/README.md]: Fixed in 83b1526

**Pull request merged**